File: src/com/fictional_particle_sim/TheCanvas.java
package com.fictional_particle_sim;

import com.fictional_particle_sim.geometrics.Point;
import com.fictional_particle_sim.geometrics.Vector;
import com.fictional_particle_sim.physicals.Barrier;

import java.awt.*;

import static com.fictional_particle_sim.Constants.PPU;

public class TheCanvas extends Canvas{


    @Override
    public void paint(Graphics g) {

        Main.computeAndDraw(g, this);

    }

    public static void drawVector(Graphics g, Vector v, boolean arrow, double l, double h) {
        g.setColor(v.color);
        g.drawLine((int)Math.round(v.start.x), (int)Math.round(v.start.y), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
        if (arrow) {
            g.drawLine((int)Math.round(v.end.x - h * (v.end.y - v.start.y) - (l * (v.end.x - v.start.x))), (int)Math.round(v.end.y + h * (v.end.x - v.start.x) - (l * (v.end.y - v.start.y))), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
            g.drawLine((int)Math.round(v.end.x + h * (v.end.y - v.start.y) - (l * (v.end.x - v.start.x))), (int)Math.round(v.end.y - h * (v.end.x - v.start.x) - (l * (v.end.y - v.start.y))), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
        }
    }
    public static void drawVector(Graphics g, Vector v, boolean arrow, double l, double h, Color color) {
        g.setColor(color);
        g.drawLine((int)Math.round(v.start.x), (int)Math.round(v.start.y), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
        if (arrow) {
            g.drawLine((int)Math.round(v.end.x - h * (v.end.y - v.start.y) - (l * (v.end.x - v.start.x))), (int)Math.round(v.end.y + h * (v.end.x - v.start.x) - (l * (v.end.y - v.start.y))), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
            g.drawLine((int)Math.round(v.end.x + h * (v.end.y - v.start.y) - (l * (v.end.x - v.start.x))), (int)Math.round(v.end.y - h * (v.end.x - v.start.x) - (l * (v.end.y - v.start.y))), (int)Math.round(v.end.x),(int)Math.round(v.end.y));
        }
    }
    public static void drawVectors(Graphics g, Vector[] v, boolean arrow, double l, double h) {
        for (Vector vector : v) {
            drawVector(g, vector, arrow, l, h);
        }

    }
    public static void drawBarriers(Graphics g, Barrier[] v) {
        for (Barrier barrier : v) {
            g.setColor(barrier.color);
            switch (barrier.shape) {
                case "LINE": g.drawLine((int) Math.round(barrier.line.start.x * PPU), (int) Math.round(barrier.line.start.y * PPU), (int) Math.round(barrier.line.end.x * PPU), (int) Math.round(barrier.line.end.y * PPU));
                case "CIRCLE": g.drawOval((int)((barrier.center.x - barrier.radius) * PPU), (int)((barrier.center.y - barrier.radius) * PPU), (int)(2 * barrier.radius * PPU),(int)(2 * barrier.radius * PPU));
                case "RECTANGLE": g.drawRect((int)(barrier.topLeft.x * PPU), (int)(barrier.topLeft.y * PPU), (int)((barrier.bottomRight.x - barrier.topLeft.x) * PPU), (int)((barrier.bottomRight.y - barrier.topLeft.y) * PPU));
            }
        }

    }

    public static void drawPoint(Graphics g, com.fictional_particle_sim.geometrics.Point p, int r, Color color) {
        g.setColor(color);
        g.fillOval((int)Math.round(p.x)-r, (int)Math.round(p.y)-r, 2*r, 2*r);
    }
    public static void drawPoints(Graphics g, com.fictional_particle_sim.geometrics.Point[] p, int r, Color color) {
        for (Point point : p) {
            drawPoint(g, point, r, color);
        }
    }

}

File: src/com/fictional_particle_sim/util/TheCanvas.kt
package com.fictional_particle_sim.util

import com.fictional_particle_sim.Main
import com.fictional_particle_sim.geometrics.Point
import com.fictional_particle_sim.geometrics.Vector
import com.fictional_particle_sim.physicals.Barrier
import com.fictional_particle_sim.physicals.Field
import com.fictional_particle_sim.physicals.Shape
import com.fictional_particle_sim.util.Constants.Companion.GRAPHICS
import com.fictional_particle_sim.util.Graphics.*
import java.awt.Canvas
import java.awt.Color
import java.awt.Graphics
import kotlin.math.roundToInt

class TheCanvas : Canvas() {

    override fun paint(g: Graphics) {
        Main.computeAndDraw(g)
    }

    companion object {
        fun drawVector(g: Graphics, v: Vector, arrow: Boolean, l: Double, h: Double) {
            g.color = v.color
            g.drawLine(v.start.x.roundToInt(), v.start.y.roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
            if (arrow) {
                g.drawLine((v.end.x - h * (v.end.y - v.start.y) - l * (v.end.x - v.start.x)).roundToInt(), (v.end.y + h * (v.end.x - v.start.x) - l * (v.end.y - v.start.y)).roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
                g.drawLine((v.end.x + h * (v.end.y - v.start.y) - l * (v.end.x - v.start.x)).roundToInt(), (v.end.y - h * (v.end.x - v.start.x) - l * (v.end.y - v.start.y)).roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
            }
        }

        fun drawVector(g: Graphics, v: Vector, arrow: Boolean, l: Double, h: Double, color: Color) {
            if (color.alpha != 0){
                g.color = color
                g.drawLine(v.start.x.roundToInt(), v.start.y.roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
                if (arrow) {
                    g.drawLine((v.end.x - h * (v.end.y - v.start.y) - l * (v.end.x - v.start.x)).roundToInt(), (v.end.y + h * (v.end.x - v.start.x) - l * (v.end.y - v.start.y)).roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
                    g.drawLine((v.end.x + h * (v.end.y - v.start.y) - l * (v.end.x - v.start.x)).roundToInt(), (v.end.y - h * (v.end.x - v.start.x) - l * (v.end.y - v.start.y)).roundToInt(), v.end.x.roundToInt(), v.end.y.roundToInt())
                }
            }
        }

        fun drawBarriers(g: Graphics, v: Array<Barrier>) {
            for (barrier in v) {
                if (barrier.color.alpha != 0) {
                    g.color = when (GRAPHICS){
                        FAST -> Color(barrier.color.red * barrier.color.alpha/255, barrier.color.green * barrier.color.alpha/255, barrier.color.blue * barrier.color.alpha/255)
                        FANCY -> barrier.color
                    }

                    when (barrier.shape) {
                        Shape.LINE -> {
                            g.drawLine((barrier.line.start.x * Constants.PPU).roundToInt(), (barrier.line.start.y * Constants.PPU).roundToInt(), (barrier.line.end.x * Constants.PPU).roundToInt(), (barrier.line.end.y * Constants.PPU).roundToInt())
                            g.drawOval(((barrier.center.x - barrier.radius) * Constants.PPU).toInt(), ((barrier.center.y - barrier.radius) * Constants.PPU).toInt(), (2 * barrier.radius * Constants.PPU).toInt(), (2 * barrier.radius * Constants.PPU).toInt())
                            g.drawRect((barrier.topLeft.x * Constants.PPU).toInt(), (barrier.topLeft.y * Constants.PPU).toInt(), ((barrier.bottomRight.x - barrier.topLeft.x) * Constants.PPU).toInt(), ((barrier.bottomRight.y - barrier.topLeft.y) * Constants.PPU).toInt())
                        }
                        Shape.CIRCLE -> {
                            g.drawOval(((barrier.center.x - barrier.radius) * Constants.PPU).toInt(), ((barrier.center.y - barrier.radius) * Constants.PPU).toInt(), (2 * barrier.radius * Constants.PPU).toInt(), (2 * barrier.radius * Constants.PPU).toInt())
                            g.drawRect((barrier.topLeft.x * Constants.PPU).toInt(), (barrier.topLeft.y * Constants.PPU).toInt(), ((barrier.bottomRight.x - barrier.topLeft.x) * Constants.PPU).toInt(), ((barrier.bottomRight.y - barrier.topLeft.y) * Constants.PPU).toInt())
                        }
                        Shape.RECTANGLE -> g.drawRect((barrier.topLeft.x * Constants.PPU).toInt(), (barrier.topLeft.y * Constants.PPU).toInt(), ((barrier.bottomRight.x - barrier.topLeft.x) * Constants.PPU).toInt(), ((barrier.bottomRight.y - barrier.topLeft.y) * Constants.PPU).toInt())
                    }
                }
            }
        }

        fun drawFields(g: Graphics, v: Array<Field>) {
            for (field in v) {
                if (field.color.alpha != 0){
                    g.color = when (GRAPHICS){
                        FAST -> Color(field.color.red * field.color.alpha/255, field.color.green * field.color.alpha/255, field.color.blue * field.color.alpha/255)
                        FANCY -> field.color
                    }
                    when (field.shape) {
                        Shape.CIRCLE -> {
                            g.fillOval(((field.center.x - field.radius) * Constants.PPU).toInt(), ((field.center.y - field.radius) * Constants.PPU).toInt(), (2 * field.radius * Constants.PPU).toInt(), (2 * field.radius * Constants.PPU).toInt())
                            g.fillRect((field.topLeft.x * Constants.PPU).toInt(), (field.topLeft.y * Constants.PPU).toInt(), ((field.bottomRight.x - field.topLeft.x) * Constants.PPU).toInt(), ((field.bottomRight.y - field.topLeft.y) * Constants.PPU).toInt())
                        }
                        Shape.RECTANGLE -> g.fillRect((field.topLeft.x * Constants.PPU).toInt(), (field.topLeft.y * Constants.PPU).toInt(), ((field.bottomRight.x - field.topLeft.x) * Constants.PPU).toInt(), ((field.bottomRight.y - field.topLeft.y) * Constants.PPU).toInt())
                        else -> {}
                    }
                }
            }
        }

        fun drawPoint(g: Graphics, p: Point, r: Int, color: Color) {
            if (color.alpha != 0) {
                g.color = color
                g.fillOval(p.x.roundToInt() - r, p.y.roundToInt() - r, 2 * r, 2 * r)
            }
        }
        fun drawBoxAroundPoint(g: Graphics, p: Point, r: Int, color: Color) {
            if (color.alpha != 0) {
                g.color = color
                g.fillRect(p.x.roundToInt() - r, p.y.roundToInt() - r, 2 * r, 2 * r)
            }
        }
    }
}